cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -241,10 +263,123 @@ private static ReportGenerationRequest getReportRequest(Location reportIdDir) th
+                                                    ReportGenerationRequest reportRequest) throws Exception {
       // Get the creation time from the report ID, which is time based UUID
-      return new ReportMetaInfo(reportRequest.getName(), null, ReportIds.getTime(reportId, TimeUnit.SECONDS),
-                                null, getReportStatus(reportIdDir));
+      long creationTime = ReportIds.getTime(reportId, TimeUnit.SECONDS);
+      ReportStatus status = getReportStatus(reportIdDir);
+      Location savedFile = reportIdDir.append(SAVED_FILE);
+      // if the report is not saved, get the name of the report from the report generation request and include the
+      // expiry time in the response if the report is COMPLETED
+      if (!savedFile.exists()) {
+        Long expiryTimeSecs = null;
+        if (ReportStatus.COMPLETED.equals(status)) {
+          expiryTimeSecs = TimeUnit.MILLISECONDS.toSeconds(getExpiryTimeMillis(reportIdDir));
+        }
+        return new ReportMetaInfo(reportRequest.getName(), null, creationTime, expiryTimeSecs, status);
+      }
+      // read the report save request from the file
+      String reportSaveRequestString =
+        new String(ByteStreams.toByteArray(savedFile.getInputStream()), StandardCharsets.UTF_8);
+      ReportSaveRequest saveRequest = GSON.fromJson(reportSaveRequestString, REPORT_SAVE_REQUEST_TYPE);
+      // the report is saved, get the name and the description from the report save request
+      // and return null as expiry time
+      return new ReportMetaInfo(saveRequest.getName(), saveRequest.getDescription(), creationTime, null, status);
+    }
+
+    @DELETE
+    @Path("/reports/{report-id}")
+    public void deleteReport(HttpServiceRequest request, HttpServiceResponder responder,
+                             @PathParam("report-id") String reportId) throws IOException {
+      Location reportIdDir = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(reportId);
+      if (!reportIdDir.exists()) {
+        responder.sendError(404, String.format("Report with id %s does not exist.", reportId));
+        return;
+      }
+      ReportStatus status;
+      try {
+        status = getReportStatus(reportIdDir);
+      } catch (Exception e) {
+        LOG.error("Failed to get the status of the report {}", reportId, e);
+        responder.sendError(500, String.format("Failed to get the status of the report %s becasue of error: %s",
+                                               reportId, e.getMessage()));
+        return;
+      }
+      // if the report is already deleted, deleting the report again is not allowed
+      if (ReportStatus.DELETED.equals(status)) {
+        responder.sendError(404, String.format("Report with id %s has already been deleted", reportId));
+        return;
+      }
+      Location deletedFile = reportIdDir.append(TO_BE_DELETED_FILE);
+      if (!deletedFile.createNew()) {
+        reportIdDir.delete();
```

Comment:
why are you deleting the report dir if you cant create deleted file. that doesn't sound correct.